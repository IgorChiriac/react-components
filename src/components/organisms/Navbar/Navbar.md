### Summary

`<Navbar />` doesn't render anything: it's just a namespace for `<Navbar.Dropdwn />`, `<Navbar.Dropdown />` and `<Navbar.Link />`.

In order to create a navigation bar to be used within a **Zopa application**, you will need to leverage these components together.

### Examples

- Static

```js { "props": { "style": { "transform": "translate3d(0, 0, 0)", "border": "2px solid #efefef" } } }
import { Navbar, Link, HamburgerIcon, ProfileIcon, colors } from '@zopauk/react-components';

<Navbar.Layout
  backgroundColor={colors.base.primary}
  left={
    <Navbar.Dropdown
      id="unique-id"
      ariaLabel="Awesome navbar dropdown"
      items={[{ label: 'one', href: '#' }, { label: 'two', href: '#' }, { label: 'three', href: '#' }]}
      renderOpener={({ open, getOpenerProps }) => (
        <div style={{ padding: 8 }}>
          <Navbar.Link href="#" open={open} {...getOpenerProps()}>
            <HamburgerIcon size="30px" activeColor={colors.neutral.white} inactiveColor={colors.neutral.white} />
          </Navbar.Link>
        </div>
      )}
      renderItem={({ item: { label, href }, getItemProps }) => (
        <Link href={href} {...getItemProps()}>
          {label}
        </Link>
      )}
    />
  }
  right={
    <Navbar.Dropdown
      id="another-unique-id"
      ariaLabel="Awesome navbar dropdown"
      items={[{ label: 'one', href: '#' }, { label: 'two', href: '#' }]}
      renderOpener={({ open, getOpenerProps }) => {
        return (
          <div style={{ padding: 8 }}>
            <Navbar.Link color="#fff" href="#" open={open} {...getOpenerProps()}>
              <ProfileIcon activeColor="#fff" inactiveColor="#fff" />
            </Navbar.Link>
          </div>
        );
      }}
      renderItem={({ item: { label, href }, getItemProps }) => (
        <Link href={href} {...getItemProps()}>
          {label}
        </Link>
      )}
    />
  }
/>;
```

- Responsive

```js { "props": { "style": { "transform": "translate3d(0, 0, 0)", "border": "2px solid #efefef" } } }
import { Navbar, Link, HamburgerIcon, FlexCol, colors } from '@zopauk/react-components';

class ResponsiveNavbar extends React.Component {
  constructor(props) {
    super(props);
    this.renderLeft = this.renderLeft.bind(this);
  }

  renderLeft() {
    const items = [{ label: 'one', href: '#' }, { label: 'two', href: '#' }, { label: 'three', href: '#' }];
    return (
      <div>
        <FlexCol xs="auto" m="hidden">
          <Navbar.Dropdown
            id="yet-another-unique-id"
            ariaLabel="Awesome navbar dropdown"
            items={items}
            renderOpener={({ open, getOpenerProps }) => (
              <div style={{ padding: 8 }}>
                <Navbar.Link color={colors.neutral.white} href="#" open={open} {...getOpenerProps()}>
                  <HamburgerIcon size="30px" activeColor={colors.neutral.white} inactiveColor={colors.neutral.white} />
                </Navbar.Link>
              </div>
            )}
            renderItem={({ item: { label, href }, getItemProps }) => (
              <Link href={href} {...getItemProps()}>
                {label}
              </Link>
            )}
          />
        </FlexCol>
        <FlexCol xs="hidden" m="auto">
          {items.map(({ label, href }, index) => (
            <Link key={`desktop-link-${index}`} color={colors.neutral.white} href={href} style={{ padding: 8 }}>
              {label}
            </Link>
          ))}
        </FlexCol>
      </div>
    );
  }

  render() {
    return <Navbar.Layout backgroundColor="#00B9A7" left={this.renderLeft()} />;
  }
}

<ResponsiveNavbar />;
```
