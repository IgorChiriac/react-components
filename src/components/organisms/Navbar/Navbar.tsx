import React from 'react';

import NavbarDropdown from './NavbarDropdown/NavbarDropdown';
import NavbarLayout from './NavbarLayout/NavbarLayout';
import NavbarLink from './NavbarLink/NavbarLink';

class Navbar extends React.Component {
  public static Layout = NavbarLayout;
  public static Dropdown = NavbarDropdown;
  public static Link = NavbarLink;

  public render() {
    return null;
  }
}

export default Navbar;
