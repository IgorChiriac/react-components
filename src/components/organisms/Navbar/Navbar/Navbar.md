### Summary

By default [`<Navbar.LinksList />`](#/Components/Organisms/Navbar/NavbarLinksList) will render [`<Navbar.Dropdown />`](#/Components/Organisms/Navbar/NavbarDropdown) or [`<Navbar.Link />`](#/Components/Organisms/Navbar/NavbarLink) based on whether links have `children` or not.

⚠️ The menu animates when scrolling. For a better preview view [Default example](#/Components/Organisms/Navbar/Navbar/1)

### Examples

- Default theme

```ts { "props": { "style": { "transform": "translate3d(0, 0, 0)", "backgroundColor": "#FFFFFF", "overflow": "hidden", "height": "500px" } } }
import { Navbar } from '@zopauk/react-components';

const NAV_ITEMS = [
  {
    label: 'Borrow',
    href: '/loans',
    onClick: () => alert('testing2'),
    qadata: 'Borrow.topBar.Button',
    children: [
      {
        label: 'Loans',
        href: '/loans',
        onClick: () => alert('testing'),
        qadata: 'loans.Borrow.topBar.Menu',
      },
      {
        label: 'Car finance',
        groupTitle: 'carFinanceGroup',
        groupChildren: [
          {
            label: 'Car loan',
            href: '/loans/car-loans',
            qadata: 'Car_loan.Borrow.topBar.Menu',
          },
        ],
      },
      {
        label: 'Tools',
        groupTitle: 'toolsGroup',
        groupChildren: [
          {
            label: 'Borrowing power',
            href: '/borrowing-power',
            qadata: 'Borrowing_Power.Borrow.topBar.Menu',
          },
        ],
      },
    ],
  },
  {
    label: 'Invest',
    href: '/invest',
    qadata: 'Invest.topBar.Menu',
    children: [
      {
        label: 'Peer-to peer investments',
        href: '/invest',
        qadata: 'Peer_to_peer.Invest.topBar.Menu',
      },
      {
        label: 'Innovative Finance ISA',
        href: '/invest/isa',
        qadata: 'Innovative.Invest.topBar.Menu',
      },
    ],
  },
  {
    label: 'About',
    href: '/about',
    qadata: 'About.topBar.Menu',
    children: [
      {
        label: 'About Zopa',
        href: '/about',
        qadata: 'About.About.topBar.Menu',
      },
      {
        label: 'Our story',
        href: '/about/our-story',
        qadata: 'Story.About.topBar.Menu',
      },
      {
        label: 'Our board',
        href: '/about/board',
        qadata: 'Our_Board.About.topBar.Menu',
      },
      {
        label: 'Our leadership team',
        href: '/about/leadership',
        qadata: 'Our_Leadership.About.topBar.Menu',
      },
      {
        label: 'Awards',
        href: '/about/awards',
        qadata: 'Awards.About.topBar.Menu',
      },
      {
        label: 'Careers',
        href: '/about/careers',
        qadata: 'Careers.About.topBar.Menu',
      },
      {
        label: 'Press office',
        href: '/about/press',
        qadata: 'Press_office.About.topBar.Menu',
      },
    ],
  },
  {
    label: 'Support',
    href: '/contact',
    qadata: 'Support.topBar.Menu',
  },
];

<Navbar overlayLogoWith={<span></span>}>
  <Navbar.LinksList links={NAV_ITEMS} />
</Navbar>;
```

- Custom Navbar components

```ts { "props": { "style": { "transform": "translate3d(0, 0, 0)", "backgroundColor": "#FFFFFF", "overflow": "hidden", "height": "500px" } } }
import { Navbar, Link } from '@zopauk/react-components';
const NAV_ITEMS = [
  {
    label: 'Borrow',
    href: '/loans',
    qadata: 'Borrow.topBar.Button',
    children: [
      {
        label: 'Loans',
        href: '/loans',
        qadata: 'loans.Borrow.topBar.Menu',
      },
      {
        label: 'Car finance',
        groupTitle: 'carFinanceGroup',
        groupChildren: [
          {
            label: 'Car loan',
            href: '/loans/car-loans',
            qadata: 'Car_loan.Borrow.topBar.Menu',
          },
        ],
      },
      {
        label: 'Tools',
        groupTitle: 'toolsGroup',
        groupChildren: [
          {
            label: 'Borrowing power',
            href: '/borrowing-power',
            qadata: 'Borrowing_Power.Borrow.topBar.Menu',
          },
        ],
      },
    ],
  },
  {
    label: 'Invest',
    href: '/invest',
    qadata: 'Invest.topBar.Menu',
    children: [
      {
        label: 'Peer-to peer investments',
        href: '/invest',
        qadata: 'Peer_to_peer.Invest.topBar.Menu',
      },
      {
        label: 'Innovative Finance ISA',
        href: '/invest/isa',
        qadata: 'Innovative.Invest.topBar.Menu',
      },
    ],
  },
  {
    label: 'About',
    href: '/about',
    qadata: 'About.topBar.Menu',
    children: [
      {
        label: 'About Zopa',
        href: '/about',
        qadata: 'About.About.topBar.Menu',
      },
      {
        label: 'Our story',
        href: '/about/our-story',
        qadata: 'Story.About.topBar.Menu',
      },
      {
        label: 'Our board',
        href: '/about/board',
        qadata: 'Our_Board.About.topBar.Menu',
      },
      {
        label: 'Our leadership team',
        href: '/about/leadership',
        qadata: 'Our_Leadership.About.topBar.Menu',
      },
      {
        label: 'Awards',
        href: '/about/awards',
        qadata: 'Awards.About.topBar.Menu',
      },
      {
        label: 'Careers',
        href: '/about/careers',
        qadata: 'Careers.About.topBar.Menu',
      },
      {
        label: 'Press office',
        href: '/about/press',
        qadata: 'Press_office.About.topBar.Menu',
      },
    ],
  },
  {
    label: 'Support',
    href: '/contact',
    qadata: 'Support.topBar.Menu',
  },
];

<Navbar overlayLogoWith={<a href="https://www.zopa.com" />}>
  <Navbar.LinksList
    links={NAV_ITEMS}
    renderLink={(item, props) => (
      <Link href={item.href} target="_blank" {...props} className="px-4 py-2">
        {item.label}
      </Link>
    )}
  />
</Navbar>;
```

- Without CTA

```ts { "props": { "style": { "transform": "translate3d(0, 0, 0)", "backgroundColor": "#FFFFFF", "overflow": "hidden", "height": "500px" } } }
import { Navbar, Link, buttonStyle } from '@zopauk/react-components';
import styled from 'styled-components';

const NAV_ITEMS = [
  {
    label: 'Borrow',
    href: '/loans',
    qadata: 'Borrow.topBar.Button',
    children: [
      {
        label: 'Loans',
        href: '/loans',
        qadata: 'loans.Borrow.topBar.Menu',
      },
      {
        label: 'Car finance',
        groupTitle: 'carFinanceGroup',
        groupChildren: [
          {
            label: 'Car loan',
            href: '/loans/car-loans',
            qadata: 'Car_loan.Borrow.topBar.Menu',
          },
        ],
      },
      {
        label: 'Tools',
        groupTitle: 'toolsGroup',
        groupChildren: [
          {
            label: 'Borrowing power',
            href: '/borrowing-power',
            qadata: 'Borrowing_Power.Borrow.topBar.Menu',
          },
        ],
      },
    ],
  },
  {
    label: 'Invest',
    href: '/invest',
    qadata: 'Invest.topBar.Menu',
    children: [
      {
        label: 'Peer-to peer investments',
        href: '/invest',
        qadata: 'Peer_to_peer.Invest.topBar.Menu',
      },
      {
        label: 'Innovative Finance ISA',
        href: '/invest/isa',
        qadata: 'Innovative.Invest.topBar.Menu',
      },
    ],
  },
  {
    label: 'About',
    href: '/about',
    qadata: 'About.topBar.Menu',
    children: [
      {
        label: 'About Zopa',
        href: '/about',
        qadata: 'About.About.topBar.Menu',
      },
      {
        label: 'Our story',
        href: '/about/our-story',
        qadata: 'Story.About.topBar.Menu',
      },
      {
        label: 'Our board',
        href: '/about/board',
        qadata: 'Our_Board.About.topBar.Menu',
      },
      {
        label: 'Our leadership team',
        href: '/about/leadership',
        qadata: 'Our_Leadership.About.topBar.Menu',
      },
      {
        label: 'Awards',
        href: '/about/awards',
        qadata: 'Awards.About.topBar.Menu',
      },
      {
        label: 'Careers',
        href: '/about/careers',
        qadata: 'Careers.About.topBar.Menu',
      },
      {
        label: 'Press office',
        href: '/about/press',
        qadata: 'Press_office.About.topBar.Menu',
      },
    ],
  },
  {
    label: 'Support',
    href: '/contact',
    qadata: 'Support.topBar.Menu',
  },
];

const ButtonLink = styled(Link)`
  ${buttonStyle}
  margin-left: 8px;
`;

<Navbar withCTA={false}>
  <Navbar.LinksList links={NAV_ITEMS} />
</Navbar>;
```

- Without links

```ts { "props": { "style": { "transform": "translate3d(0, 0, 0)", "backgroundColor": "#FFFFFF", "overflow": "hidden", "height": "500px" } } }
import { Navbar, Link, buttonStyle } from '@zopauk/react-components';
import styled from 'styled-components';

const ButtonLink = styled(Link)`
  ${buttonStyle}
  margin-left: 8px;
`;

<Navbar
  overlayLogoWith={<a href="https://www.zopa.com" />}
  cta={
    <ButtonLink href="https://www.zopa.com" styling="primary">
      Sign-in
    </ButtonLink>
  }
></Navbar>;
```
