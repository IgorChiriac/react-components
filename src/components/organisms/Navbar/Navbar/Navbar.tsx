import React from 'react';
import styled from 'styled-components';

import { colors } from '../../../../constants/colors';
import { navbarHeight } from '../../../../constants/components';
import useScrollThreshold from '../useScrollThreshold/useScrollThreshold';
import FlexContainer from '../../../layout/FlexContainer/FlexContainer';

export interface ILayoutOuterProps extends React.HTMLAttributes<HTMLDivElement> {
  backgroundColor: string;
  overlap: boolean;
}

const LayoutOuter = styled.div<ILayoutOuterProps>`
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  z-index: 1;
  background-color: ${({ backgroundColor }) => backgroundColor};
  height: ${navbarHeight}px;
  transition: box-shadow 0.2s ease-in-out;
  box-shadow: 0 0 0 transparent;
  ${({ overlap }) => overlap && `box-shadow: rgba(0, 0, 0, 0.2) 0 1px 2px;`}
`;

const LayoutInner = styled.nav`
  display: flex;
  align-items: center;
  justify-content: space-between;
  max-width: 960px;
  height: 100%;
  width: 100%;
  margin-left: auto;
  margin-right: auto;
`;

const Spacer = styled.div`
  height: ${navbarHeight}px;
`;

const Container = styled(FlexContainer)`
  height: inherit;
`;

export interface INavbarProps {
  /**
   * background color
   */
  backgroundColor?: string;
  /**
   * Component appearing on the left
   */
  left?: React.ReactNode;
  /**
   * Component appearing in the center
   */
  center?: React.ReactNode;
  /**
   * Component appearing on the right
   */
  right?: React.ReactNode;
}

const Navbar = ({ backgroundColor = colors.neutral.white, left, center, right }: INavbarProps) => {
  const overThreshold = useScrollThreshold();

  return (
    <>
      <Spacer />
      <LayoutOuter backgroundColor={backgroundColor} overlap={overThreshold}>
        <Container>
          <LayoutInner>
            {left && <div>{left}</div>}
            {center && <div>{center}</div>}
            {right && <div>{right}</div>}
          </LayoutInner>
        </Container>
      </LayoutOuter>
    </>
  );
};

export default Navbar;
