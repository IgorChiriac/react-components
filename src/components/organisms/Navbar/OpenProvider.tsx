import React, { createContext, Dispatch, FC, SetStateAction, useState } from 'react';

interface OpenContextType {
  open: boolean;
  setOpen: Dispatch<SetStateAction<boolean>>;
}

const OpenContext = createContext<OpenContextType>({
  open: false,
  setOpen: () => {
    /* Empty method to enable typing for useState */
  },
});

const OpenProvider: FC = ({ children }) => {
  const [open, setOpen] = useState(false);

  return <OpenContext.Provider value={{ open, setOpen }}>{children}</OpenContext.Provider>;
};

const useOpen = () => React.useContext(OpenContext);

export { OpenProvider, useOpen };
