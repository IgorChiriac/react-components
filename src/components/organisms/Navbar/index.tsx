import React, { FC } from 'react';

import Navbar, { NavbarProps } from './Navbar/Navbar';
import NavbarAction from './NavbarAction/NavbarAction';
import { OpenProvider } from './OpenProvider';

interface NavbarStatic {
  Action: typeof NavbarAction;
}

const NavbarWrapper: NavbarStatic & FC<NavbarProps> = (props) => (
  <OpenProvider>
    <Navbar {...props} />
  </OpenProvider>
);

NavbarWrapper.Action = NavbarAction;

export default NavbarWrapper;
export { navbarLinkStyles } from './NavbarLink/NavbarLink';
