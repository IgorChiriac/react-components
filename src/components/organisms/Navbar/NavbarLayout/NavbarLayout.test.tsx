import React from 'react';
import { render } from 'react-testing-library';
import Navbar from '../Navbar';

describe('<Navbar.Layout />', () => {
  it('should render component with default props', () => {
    const { container } = render(<Navbar.Layout />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render component with all the props', () => {
    const left = <span>left</span>;
    const center = <span>center</span>;
    const right = <span>right</span>;
    const { container } = render(<Navbar.Layout left={left} center={center} right={right} />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
