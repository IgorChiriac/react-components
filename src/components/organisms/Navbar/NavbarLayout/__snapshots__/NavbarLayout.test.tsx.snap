// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Navbar.Layout /> should render component with all the props 1`] = `
.c0 {
  height: 80px;
}

<div
  class="c0"
/>
`;

exports[`<Navbar.Layout /> should render component with default props 1`] = `
.c0 {
  height: 80px;
}

<div
  class="c0"
/>
`;
